### ISSUE TITLE
XF app using location services when background is terminated on IOS13

### ISSUE THREAD

**Issue by Chris-Marassovich** [xamarin/xamarin-macios#6827]
[IOSLocationTest.zip](https://github.com/xamarin/xamarin-macios/files/3527952/IOSLocationTest.zip)
### Steps to Reproduce

1. Open sample repro solution OR easily create one yourself by .. new XF app master/detail, IOS.
2. copy location services implementation from https://docs.microsoft.com/en-au/xamarin/ios/app-fundamentals/backgrounding/ios-backgrounding-walkthroughs/location-walkthrough#application-set-up
3. Run up app on IOS13 devices .. background app and after a few minutes it will be terminated.

### Expected Behavior
App will happily stay alive and receive location updates (if any) while in background.  It does on IOS12 and lower.
### Actual Behavior
Terminated by IOS after a few minutes (approx. 2 minutes)
### Environment
IOS13 device.
1.
Microsoft Visual Studio Professional 2019
Version 16.2.3
VisualStudio.16.Release/16.2.3+29215.179
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Professional

Application Insights Tools for Visual Studio Package   9.1.00611.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2019   16.2.290.48383
ASP.NET and Web Tools 2019

ASP.NET Web Frameworks and Tools 2019   16.2.290.48383
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.2.290.48383
Azure App Service Tools v3.0.0

Azure Functions and Web Jobs Tools   16.2.290.48383
Azure Functions and Web Jobs Tools

C# Tools   3.2.1-beta4-19408-03+2fc6a04980f800c59e8ede97e6ae294ff47d666e
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.2.0 (d16-2@8b56e20)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 0x10 - v2.9.20626.2

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Mono Debugging for Visual Studio   16.2.6 (4cfc7c3)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   5.2.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.61906.28070
Microsoft SQL Server Data Tools

TypeScript Tools   16.0.10627.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.2.1-beta4-19408-03+2fc6a04980f800c59e8ede97e6ae294ff47d666e
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.4 for F# 4.6   16.2.0-beta.19321.1+a24d94ecf97d0d69d4fbe6b8b10cd1f97737fff4
Microsoft Visual F# Tools 10.4 for F# 4.6

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Spell Check Everywhere   VSSpellCheckEverywhere
An extension that enables spell checking within any Visual Studio file editor or tool window that uses WPF text boxes.
https://GitHub.com/EWSoftware/VSSpellChecker

Visual Studio Spell Checker   VSSpellChecker
An editor extension that checks the spelling of comments, strings, and plain text as you type or interactively with tool windows.
https://GitHub.com/EWSoftware/VSSpellChecker

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   16.2.0.95 (d16-2@37df81894)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   16.2.0.375 (remotes/origin/d16-2@357d38ef4)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   16.3.117 (59a59e8)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.4.1.0 (d16-2/ebd65c9)
Xamarin.Android Reference Assemblies and MSBuild support.
    Mono: mono/mono/2019-02@e6f5369c2d2
    Java.Interop: xamarin/java.interop/d16-2@d64ada5
    LibZipSharp: grendello/LibZipSharp/d16-2@caa0c74
    LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
    ProGuard: xamarin/proguard/master@905836d
    SQLite: xamarin/sqlite/3.27.1@8212a2d
    Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-2@6f6c969

Xamarin.iOS and Xamarin.Mac SDK   12.14.0.114 (c669116)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.

### Build Logs
Not a build issue
### Example Project (If Possible)
One is attached.

**Comment by rolfbjarne** [comment]
Thank you for your feedback!

I tried running the sample project on a device with iOS 13 beta 6, and it's been running fine for over half an hour now, without any crashes.

Could you attach the Application Output from a run that crashes, as well as any crash reports?

We look forward to hearing from you!

**Comment by Chris-Marassovich** [comment]
Thanks rolfbjarne for getting back so quickly.
Just a few things I want to confirm first as I have tried the attached zip on two separate IOS devices running IOS13 and they both show the same issue.
1. Are you sending the app to the background?  It certainly runs fine when foregrounded. The issue is when backgrounding the app.
2. I am not using any of the preview or beta versions for IOS 13 such as VS or the SDKs.  I wonder if that could be the issue?  I am running all the current production versions of VS and Xamarin.

At the moment I am not on site so I cannot post the output from the console logs but will be back on site in a day or so.

Cheers
Chris ...

**Comment by rolfbjarne** [comment]
I didn't actually send the app to the background, but I tested again, and it still works fine for me. The first time the notifications stopped, but when I opened the app again I got a dialog asking if I wanted to allow the app to always receive notifications or only when in use, I chose always, and after that the app started getting notifications in the background too (and it doesn't crash).

I'm testing with the stable versions of Xamarin.iOS, so the version shouldn't be an issue, but you can of course try the preview version of our Xcode 11 support and see how that goes.

**Comment by Chris-Marassovich** [comment]
Hi Rolf.
I am using VS 16.2.3 and I noted that Xcode was v10.2 so I updated that to the latest which is v 10.3.
Built and deployed the app.  It was getting location updates in the background for 10 minutes before being terminated.  Subsequent runs are now being terminated within 2 minutes after entering the background.
This is very strange as the same app on almost identical env (yours and mine) is not functioning the same.

Can you please confirm the versions of VS and Xcode you built with?

I start the app and as soon as possible I background it.  That's around line 260 it enters the background.
It continues to get location updates for a period then around line 440 they are no longer delivered which makes sense as the device is not moving.
At 492 it is terminated.

Can you shed any light on why?

Cheers
Chris ...
[console.zip](https://github.com/xamarin/xamarin-macios/files/3539122/console.zip)


I have also attached the console output from my run for your review.

**Comment by Chris-Marassovich** [comment]
Hi Rolf.
I have noted another change in behaviour based on how I interact with that app.
I am on windows and the device is plugged into the Mac.
My tests so far have followed this process.
1. Start app and almost immediately place it into the background.  I bg the app via the home button.
2. I also often lock the screen or if I don't the lock period is very short on this device.
3. some seconds later I note my app stops receiving location updates and some seconds later is terminated.

What I have just done was a little different.
1. Open app and almost immediately placed it into the BG via the home button.  BUT I kept using other apps ..this kept the phone awake.  I noted that my app was still receiving location updates.
2. Lock the phone and I note that my app ceases getting location updates.
3. unlock the phone and I note my app started to get location updates.
4. I repeated steps 2 and 3 over a 10 minute period and the same actions occur and the app is yet to be terminated.
5. I locked the phone and while my app has ceased getting location updates it has yet to be terminated.  At this point it has been in the bg for over 40 minutes.

It seems if you open the app, background the app and lock the phone all while the app is subscribed to locations services (fore or background) it will likely be terminated by IOS within a few minutes.

If you background the app quickly but keep the phone awake for some period all works as expected.

Very strange.
I wonder if you follow my first test steps (open, background and lock the phone quickly) will you see what I see and the app is terminated within a few minutes.  Proof of that is in previous post with console logs.

Cheers
Chris ....

**Comment by rolfbjarne** [comment]
The problem is that the app is terminated because of high CPU usage:

> IOSLocationTest.iOS[267] used 48.00s of CPU over 57.83 seconds (averaging 83%), violating a CPU usage limit of 48.00s over 60 seconds.

My guess is that this is because iOS terminates the TCP connection between the app and the IDE when you lock the device, so the app starts using more CPU to try to connect to the IDE again, and that eventually triggers the cpu watchdog.

You could test this theory by launching the app by tapping on it, instead of launching it from the IDE, in which case the debugger won't be running. If I'm right, your app won't crash anymore. If it still crashes, please attach an updated device log like the you've already attached.

**Comment by Chris-Marassovich** [comment]
The IDE was never in play during the testing. I deployed from VS but I always ran the app via a tap once deployed.

I did that from the start as I didn't need debugging.

Yes I saw the message about excessive CPU yet there is not much going on in the code and you also built and run the app and you didn't see the same?

**Comment by rolfbjarne** [comment]
> Yes I saw the message about excessive CPU yet there is not much going on in the code and you also built and run the app and you didn't see the same?

No, I didn't, but it might have been because I used a faster device, which didn't trigger the watchdog. Which device do you use?

**Comment by Chris-Marassovich** [comment]
I am using a fast device also.
I am using an iPhone 8 or if I have the number wrong it is the model prior to the X. I am not in the office.

Do you consider that a powerful device?

**Comment by rolfbjarne** [comment]
The watchdog triggered very close to the limit (it triggered at 48.00s of CPU over 57.83 when the limit was 48.00s over 60 seconds), which means that a device doesn't have to be much faster to not trigger the limit.

I have the latest iPad Pro. I'll try with an earlier device to see what happens.

**Comment by Chris-Marassovich** [comment]
I thought I would post this so it helps anyone else.
What I am seeing is some very strange behaviour by IOS13 based on how I use the app.
Specifically how quickly I do things.

Scenario 1:
This is what I did initially to see this issue and what I continued to do for some time.
I opened the app and almost immediately sent it to the background via the home button and then locked the device.
My app receives location updates for a few seconds to a minute then the updates cease and a few minutes later the app is terminated by IOS.

Scenario 2:
Open the app and send it to the background quickly via the home button.  But now I keep the phone alive by using some other app for a short period.  Once I do this regardless of locking the device my app continues to receive location updates in the background as required (when movement occurs) but the app stays up and is not terminated by IOS.  It stayed alive until the battery died.

This is too weird and obviously beyond my code and debugging so I am going to drop this until IOS 13 comes out of beta or at least the dev and build chains are no longer preview/beta versions.

Thanks for your help Rolf.
